sap/src/test/java/org/frankframework/extensions/sap/SapListenerTest.java

```diff
@@ -0,0 +1,147 @@
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+public class SapListenerTest {
+
+	private final String sapSystemName = "sap-system";
+
+	private SapListener listener;
+	private SapSystem sapSystem;
+	private Receiver receiver;
+
+	@BeforeEach
+	public void setUp() throws SapException {
+		GlobalListItem.clear();
+
+		listener = spy(new SapListener());
+		listener.setName(listener.getClass().getSimpleName()+" under test");
+
+		receiver = mock(Receiver.class);
+		listener.setHandler(receiver);
+
+		sapSystem = spy(SapSystem.class);
+		doNothing().when(sapSystem).openSystem();
+
+		sapSystem.setName(sapSystemName);
+		sapSystem.registerItem(listener);
+	}
+
+	@Test
+	public void testNoFunctionName() {
+		listener.setProgid(null);
+
+		assertThrows(ConfigurationException.class, () -> listener.configure());
+	}
+
+	@Test
+	public void testSapSystemDoesntExist() {
+		listener.setProgid("prog-id");
+		listener.setSapSystemName("doesnt-exist");
+
+		assertThrows(NullPointerException.class, () -> listener.configure());
+	}
+
+	@Test
+	public void testSapSystemExists() {
+		listener.setProgid("prog-id");
+		listener.setSapSystemName(sapSystemName);
+
+		assertDoesNotThrow(() -> listener.configure());
+	}
+
+	@Test
+	public void testExtractMessage() {
+		listener.setProgid("prog-id");
+		listener.setSapSystemName(sapSystemName);
+
+		assertDoesNotThrow(() -> listener.configure());
+
+		JCoFunction function = mock(JCoFunction.class);
+		RawMessageWrapper<JCoFunction> inputMessage = new RawMessageWrapper<>(function);
+
+		try (Message outputMessage = listener.extractMessage(inputMessage, Map.of())) {
+			assertEquals("<request function=\"null\"></request>", outputMessage.asString());
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Test
+	public void testHandleRequestFunction() throws ListenerException {
+		listener.setProgid("prog-id");
+		listener.setSapSystemName(sapSystemName);
+
+		ArgumentCaptor<RawMessageWrapper> messageCaptor = ArgumentCaptor.forClass(RawMessageWrapper.class);
+		doNothing().when(receiver).processRawMessage(any(IListener.class), messageCaptor.capture(), any(PipeLineSession.class), anyBoolean());
+
+		JCoFunction function = mock(JCoFunction.class);
+
+		assertDoesNotThrow(() -> listener.configure());
+		assertDoesNotThrow(() -> listener.handleRequest(mock(JCoServerContext.class), function));
+
```

Comment:
Kan je hier nog een listener.afterMessageProcessed(new PipeLineResult(), messageCaptor.getValue(), new PipeLineSession()); aanroepen?

Reply:
Pffff in die functie zit een `instanceof` check wat nooit op true komt, omdat ik een mocked object meegeef..